File: classified/packages/core/src/__tests__/logger.test.ts
import { describe, expect, it, beforeEach, afterEach } from 'bun:test';
import { mock } from 'bun:test';
import { createLogger, logger, elizaLogger } from '../logger';

// Mock environment variables
const mockEnv = {
  LOG_LEVEL: '',
  DEFAULT_LOG_LEVEL: '',
  LOG_JSON_FORMAT: '',
  SENTRY_LOGGING: 'false',
  LOG_DIAGNOSTIC: '',
  LOG_TIMESTAMPS: '',
};

// Mock adze
mock.module('adze', () => ({
  adze: {
    withConfig: mock(() => ({
      info: mock(),
      error: mock(),
      warn: mock(),
      debug: mock(),
      verbose: mock(),
      trace: mock(),
      custom: mock(),
    })),
    use: mock(),
  },
  createShed: mock(() => ({
    bundle: mock(() => ({})),
    clear: mock(),
  })),
}));

describe('Logger', () => {
  let originalEnv: NodeJS.ProcessEnv;

  beforeEach(() => {
    // Save original environment
    originalEnv = { ...process.env };
    // Reset environment variables
    Object.keys(mockEnv).forEach((key) => {
      process.env[key] = mockEnv[key];
    });
    mock.restore();
  });

  afterEach(() => {
    // Restore original environment
    process.env = originalEnv;
  });

  describe('Logger Creation', () => {
    it('should export logger instance', () => {
      expect(logger).toBeDefined();
      expect(typeof logger.info).toBe('function');
      expect(typeof logger.error).toBe('function');
      expect(logger.warn).toBeDefined();
      expect(logger.debug).toBeDefined();
      expect(typeof logger.warn).toBe('function');
      expect(typeof logger.debug).toBe('function');
    });

    it('should export elizaLogger as alias for backward compatibility', () => {
      expect(elizaLogger).toBeDefined();
      expect(elizaLogger).toBe(logger);
    });

    it('should have custom log levels', () => {
      const testLogger = createLogger({ agentName: 'test' });
      expect(typeof testLogger.fatal).toBe('function');
      expect(typeof testLogger.error).toBe('function');
      expect(typeof testLogger.warn).toBe('function');
      expect(typeof testLogger.info).toBe('function');
      expect(typeof testLogger.debug).toBe('function');
      expect(typeof testLogger.trace).toBe('function');
      expect(typeof testLogger.verbose).toBe('function');
    });

    it('should have custom methods (log, progress, success)', () => {
      const testLogger = createLogger();
      expect(typeof testLogger.log).toBe('function');
      expect(typeof testLogger.progress).toBe('function');
      expect(typeof testLogger.success).toBe('function');
    });
  });

  describe('createLogger Function', () => {
    it('should create logger with bindings', () => {
      const bindings = { agentName: 'TestAgent', agentId: '123' };
      const customLogger = createLogger(bindings);

      expect(customLogger).toBeDefined();
      expect(typeof customLogger.info).toBe('function');
    });

    it('should create logger without bindings', () => {
      const customLogger = createLogger();

      expect(customLogger).toBeDefined();
      expect(typeof customLogger.info).toBe('function');
    });

    it('should handle boolean bindings parameter', () => {
      const customLogger = createLogger(false);

      expect(customLogger).toBeDefined();
      expect(typeof customLogger.info).toBe('function');
    });
  });

  describe('Log Level Configuration', () => {
    it('should use debug level when LOG_LEVEL is debug', () => {
      process.env.LOG_LEVEL = 'debug';
      const customLogger = createLogger();

      // Logger should be created with debug level
      expect(customLogger.level).toBeDefined();
      expect(customLogger.level).toBe('debug');
    });

    it('should use DEFAULT_LOG_LEVEL when LOG_LEVEL is not debug', () => {
      process.env.LOG_LEVEL = '';
      process.env.DEFAULT_LOG_LEVEL = 'warn';
      const customLogger = createLogger();

      expect(customLogger.level).toBeDefined();
      expect(customLogger.level).toBe('warn');
    });

    it('should default to info level when no log level is specified', () => {
      process.env.LOG_LEVEL = '';
      process.env.DEFAULT_LOG_LEVEL = '';
      const customLogger = createLogger();

      expect(customLogger.level).toBeDefined();
      expect(customLogger.level).toBe('info');
    });
  });

  describe('JSON Format Configuration', () => {
    it('should use JSON format when LOG_JSON_FORMAT is true', () => {
      process.env.LOG_JSON_FORMAT = 'true';
      const customLogger = createLogger();

      expect(customLogger).toBeDefined();
    });

    it('should use pretty format when LOG_JSON_FORMAT is false', () => {
      process.env.LOG_JSON_FORMAT = 'false';
      const customLogger = createLogger();

      expect(customLogger).toBeDefined();
    });
  });

  describe('Log Filtering', () => {
    it('should filter service registration logs in non-debug mode', () => {
      process.env.LOG_LEVEL = 'info';
      const customLogger = createLogger();

      // These logs should be filtered in non-debug mode
      const filteredMessages = [
        'registered successfully',
        'Registering',
        'Success:',
        'linked to',
        'Started',
      ];

      // Logger is created and can handle these messages
      filteredMessages.forEach((msg) => {
        expect(() => customLogger.info({ agentName: 'test', agentId: '123' }, msg)).not.toThrow();
      });
    });

    it('should not filter service registration logs in debug mode', () => {
      process.env.LOG_LEVEL = 'debug';
      const customLogger = createLogger();

      // In debug mode, all logs should pass through
      expect(customLogger.level).toBeDefined();
      expect(customLogger.level).toBe('debug');
    });
  });

  describe('Error Handling', () => {
    it('should handle Error objects in log messages', () => {
      const customLogger = createLogger();
      const testError = new Error('Test error');

      expect(() => customLogger.error(testError)).not.toThrow();
      expect(() => customLogger.error('Message', testError)).not.toThrow();
      expect(() =>
        customLogger.error({ context: 'test' }, 'Error occurred', testError)
      ).not.toThrow();
    });

    it('should format error messages properly', () => {
      const customLogger = createLogger();
      const testError = new Error('Test error');
      testError.name = 'TestError';

      // Should handle error formatting without throwing
      expect(() => customLogger.error(testError)).not.toThrow();
    });
  });

  describe('Clear Method', () => {
    it('should have clear method', () => {
      const customLogger = createLogger();
      expect(typeof customLogger.clear).toBe('function');
      expect(() => customLogger.clear()).not.toThrow();
    });

    it('should clear logs when called', () => {
      // Access the in-memory logger
      const destination = (logger as any)[Symbol.for('pino-destination')];
      expect(destination).toBeDefined();
      expect(typeof destination.clear).toBe('function');

      // Clear should work
      expect(() => destination.clear()).not.toThrow();
    });
  });

  describe('Child Logger', () => {
    it('should create child logger with additional bindings', () => {
      const customLogger = createLogger({ parent: 'main' });
      const childLogger = customLogger.child({ child: 'sub' });

      expect(childLogger).toBeDefined();
      expect(typeof childLogger.info).toBe('function');
    });
  });

  describe('Log Input Formats', () => {
    it('should handle various log input formats', () => {
      const customLogger = createLogger();

      // Test various input formats
      expect(() => customLogger.info('Simple string')).not.toThrow();
      expect(() => customLogger.info({ key: 'value' }, 'With object')).not.toThrow();
      expect(() => customLogger.info('Multiple', 'string', 'parts')).not.toThrow();
      expect(() => customLogger.error(new Error('Test'), 'With error')).not.toThrow();
    });

    it('should handle mixed arguments with errors', () => {
      const customLogger = createLogger();
      const error = new Error('Test error');

      expect(() => customLogger.error('Message', error, { extra: 'data' })).not.toThrow();
    });
  });

  describe('Legacy Pino Compatibility', () => {
    it('should have legacy pino-compatible properties', () => {
      const customLogger = createLogger();

      expect(customLogger.level).toBeDefined();
      expect(customLogger.levels).toBeDefined();
      expect(customLogger.levels.values).toBeDefined();
      expect(typeof customLogger.isLevelEnabled).toBe('function');
      expect(typeof customLogger.flush).toBe('function');
      expect(typeof customLogger.flushSync).toBe('function');
    });

    it('should have destination accessible via symbol', () => {
      const customLogger = createLogger();
      const destination = (customLogger as any)[Symbol.for('pino-destination')];

      expect(destination).toBeDefined();
      expect(typeof destination.recentLogs).toBe('function');
      expect(typeof destination.clear).toBe('function');
    });
  });

  describe('In-Memory Logger', () => {
    it('should store recent logs', () => {
      const destination = (logger as any)[Symbol.for('pino-destination')];
      expect(destination).toBeDefined();

      const logs = destination.recentLogs();
      expect(Array.isArray(logs)).toBe(true);
    });

    it('should add logs with proper format', () => {
      const destination = (logger as any)[Symbol.for('pino-destination')];

      // Add a test log
      destination.addLog({
        time: Date.now(),
        level: 30,
        msg: 'Test message',
        agentName: 'TestAgent',
      });

      const logs = destination.recentLogs();
      const lastLog = logs[logs.length - 1];

      expect(lastLog).toBeDefined();
      expect(lastLog.msg).toBe('Test message');
      expect(lastLog.agentName).toBe('TestAgent');
    });
  });

  describe('Log Level Methods', () => {
    it('should support all log levels', () => {
      const customLogger = createLogger();

      expect(() => customLogger.fatal('Fatal message')).not.toThrow();
      expect(() => customLogger.error('Error message')).not.toThrow();
      expect(() => customLogger.warn('Warn message')).not.toThrow();
      expect(() => customLogger.info('Info message')).not.toThrow();
      expect(() => customLogger.debug('Debug message')).not.toThrow();
      expect(() => customLogger.trace('Trace message')).not.toThrow();
      expect(() => customLogger.verbose('Verbose message')).not.toThrow();
      expect(() => customLogger.log('Log message')).not.toThrow();
      expect(() => customLogger.progress('Progress message')).not.toThrow();
      expect(() => customLogger.success('Success message')).not.toThrow();
    });
  });

  describe('Timestamp Configuration', () => {
    it('should respect LOG_TIMESTAMPS setting', () => {
      process.env.LOG_TIMESTAMPS = 'false';
      const customLogger = createLogger();
      expect(customLogger).toBeDefined();
    });

    it('should default to showing timestamps', () => {
      delete process.env.LOG_TIMESTAMPS;
      const customLogger = createLogger();
      expect(customLogger).toBeDefined();
    });
  });

  describe('Sentry Integration', () => {
    it('should not capture exceptions when SENTRY_LOGGING is false', () => {
      process.env.SENTRY_LOGGING = 'false';
      const customLogger = createLogger();
      const error = new Error('Test error');

      expect(() => customLogger.error(error)).not.toThrow();
    });

    it('should handle Sentry integration when enabled', () => {
      process.env.SENTRY_LOGGING = '';
      const customLogger = createLogger();
      const error = new Error('Sentry test');

      expect(() => customLogger.error(error)).not.toThrow();
    });
  });
});

File: cpp_hybrid/logger.test.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/core/src/__tests__/logger.test.h"

object mockEnv = object{
    object::pair{std::string("LOG_LEVEL"), string_empty}, 
    object::pair{std::string("DEFAULT_LOG_LEVEL"), string_empty}, 
    object::pair{std::string("LOG_JSON_FORMAT"), string_empty}, 
    object::pair{std::string("SENTRY_LOGGING"), std::string("false")}, 
    object::pair{std::string("LOG_DIAGNOSTIC"), string_empty}, 
    object::pair{std::string("LOG_TIMESTAMPS"), string_empty}
};

void Main(void)
{
    mock->module(std::string("adze"), [=]() mutable
    {
        return (object{
            object::pair{std::string("adze"), object{
                object::pair{std::string("withConfig"), mock([=]() mutable
                {
                    return (object{
                        object::pair{std::string("info"), mock()}, 
                        object::pair{std::string("error"), mock()}, 
                        object::pair{std::string("warn"), mock()}, 
                        object::pair{std::string("debug"), mock()}, 
                        object::pair{std::string("verbose"), mock()}, 
                        object::pair{std::string("trace"), mock()}, 
                        object::pair{std::string("custom"), mock()}
                    });
                }
                )}, 
                object::pair{std::string("use"), mock()}
            }}, 
            object::pair{std::string("createShed"), mock([=]() mutable
            {
                return (object{
                    object::pair{std::string("bundle"), mock([=]() mutable
                    {
                        return (object{});
                    }
                    )}, 
                    object::pair{std::string("clear"), mock()}
                });
            }
            )}
        });
    }
    );
    describe(std::string("Logger"), [=]() mutable
    {
        shared<std::shared_ptr<NodeJS::ProcessEnv>> originalEnv;
        beforeEach([=]() mutable
        {
            originalEnv = utils::assign(object{
            }, process->env);
            Object->keys(mockEnv)->forEach([=](auto key) mutable
            {
                process->env[key] = const_(mockEnv)[key];
            }
            );
            mock->restore();
        }
        );
        afterEach([=]() mutable
        {
            process->env = originalEnv;
        }
        );
        describe(std::string("Logger Creation"), [=]() mutable
        {
            it(std::string("should export logger instance"), [=]() mutable
            {
                expect(logger)->toBeDefined();
                expect(type_of(logger["info"]))->toBe(std::string("function"));
                expect(type_of(logger["error"]))->toBe(std::string("function"));
                expect(logger["warn"])->toBeDefined();
                expect(logger["debug"])->toBeDefined();
                expect(type_of(logger["warn"]))->toBe(std::string("function"));
                expect(type_of(logger["debug"]))->toBe(std::string("function"));
            }
            );
            it(std::string("should export elizaLogger as alias for backward compatibility"), [=]() mutable
            {
                expect(elizaLogger)->toBeDefined();
                expect(elizaLogger)->toBe(logger);
            }
            );
            it(std::string("should have custom log levels"), [=]() mutable
            {
                auto testLogger = createLogger(object{
                    object::pair{std::string("agentName"), std::string("test")}
                });
                expect(type_of(testLogger["fatal"]))->toBe(std::string("function"));
                expect(type_of(testLogger["error"]))->toBe(std::string("function"));
                expect(type_of(testLogger["warn"]))->toBe(std::string("function"));
                expect(type_of(testLogger["info"]))->toBe(std::string("function"));
                expect(type_of(testLogger["debug"]))->toBe(std::string("function"));
                expect(type_of(testLogger["trace"]))->toBe(std::string("function"));
                expect(type_of(testLogger["verbose"]))->toBe(std::string("function"));
            }
            );
            it(std::string("should have custom methods (log, progress, success)"), [=]() mutable
            {
                auto testLogger = createLogger();
                expect(type_of(testLogger["log"]))->toBe(std::string("function"));
                expect(type_of(testLogger["progress"]))->toBe(std::string("function"));
                expect(type_of(testLogger["success"]))->toBe(std::string("function"));
            }
            );
        }
        );
        describe(std::string("createLogger Function"), [=]() mutable
        {
            it(std::string("should create logger with bindings"), [=]() mutable
            {
                auto bindings = object{
                    object::pair{std::string("agentName"), std::string("TestAgent")}, 
                    object::pair{std::string("agentId"), std::string("123")}
                };
                auto customLogger = createLogger(bindings);
                expect(customLogger)->toBeDefined();
                expect(type_of(customLogger["info"]))->toBe(std::string("function"));
            }
            );
            it(std::string("should create logger without bindings"), [=]() mutable
            {
                auto customLogger = createLogger();
                expect(customLogger)->toBeDefined();
                expect(type_of(customLogger["info"]))->toBe(std::string("function"));
            }
            );
            it(std::string("should handle boolean bindings parameter"), [=]() mutable
            {
                auto customLogger = createLogger(false);
                expect(customLogger)->toBeDefined();
                expect(type_of(customLogger["info"]))->toBe(std::string("function"));
            }
            );
        }
        );
        describe(std::string("Log Level Configuration"), [=]() mutable
        {
            it(std::string("should use debug level when LOG_LEVEL is debug"), [=]() mutable
            {
                process->env->LOG_LEVEL = std::string("debug");
                auto customLogger = createLogger();
                expect(customLogger["level"])->toBeDefined();
                expect(customLogger["level"])->toBe(std::string("debug"));
            }
            );
            it(std::string("should use DEFAULT_LOG_LEVEL when LOG_LEVEL is not debug"), [=]() mutable
            {
                process->env->LOG_LEVEL = string_empty;
                process->env->DEFAULT_LOG_LEVEL = std::string("warn");
                auto customLogger = createLogger();
                expect(customLogger["level"])->toBeDefined();
                expect(customLogger["level"])->toBe(std::string("warn"));
            }
            );
            it(std::string("should default to info level when no log level is specified"), [=]() mutable
            {
                process->env->LOG_LEVEL = string_empty;
                process->env->DEFAULT_LOG_LEVEL = string_empty;
                auto customLogger = createLogger();
                expect(customLogger["level"])->toBeDefined();
                expect(customLogger["level"])->toBe(std::string("info"));
            }
            );
        }
        );
        describe(std::string("JSON Format Configuration"), [=]() mutable
        {
            it(std::string("should use JSON format when LOG_JSON_FORMAT is true"), [=]() mutable
            {
                process->env->LOG_JSON_FORMAT = std::string("true");
                auto customLogger = createLogger();
                expect(customLogger)->toBeDefined();
            }
            );
            it(std::string("should use pretty format when LOG_JSON_FORMAT is false"), [=]() mutable
            {
                process->env->LOG_JSON_FORMAT = std::string("false");
                auto customLogger = createLogger();
                expect(customLogger)->toBeDefined();
            }
            );
        }
        );
        describe(std::string("Log Filtering"), [=]() mutable
        {
            it(std::string("should filter service registration logs in non-debug mode"), [=]() mutable
            {
                process->env->LOG_LEVEL = std::string("info");
                shared customLogger = createLogger();
                auto filteredMessages = array<string>{ std::string("registered successfully"), std::string("Registering"), std::string("Success:"), std::string("linked to"), std::string("Started") };
                filteredMessages->forEach([=](auto msg) mutable
                {
                    expect([=]() mutable
                    {
                        return customLogger["info"](object{
                            object::pair{std::string("agentName"), std::string("test")}, 
                            object::pair{std::string("agentId"), std::string("123")}
                        }, msg);
                    }
                    )->not->toThrow();
                }
                );
            }
            );
            it(std::string("should not filter service registration logs in debug mode"), [=]() mutable
            {
                process->env->LOG_LEVEL = std::string("debug");
                auto customLogger = createLogger();
                expect(customLogger["level"])->toBeDefined();
                expect(customLogger["level"])->toBe(std::string("debug"));
            }
            );
        }
        );
        describe(std::string("Error Handling"), [=]() mutable
        {
            it(std::string("should handle Error objects in log messages"), [=]() mutable
            {
                shared customLogger = createLogger();
                shared testError = std::make_shared<Error>(std::string("Test error"));
                expect([=]() mutable
                {
                    return customLogger["error"](testError);
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["error"](std::string("Message"), testError);
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["error"](object{
                        object::pair{std::string("context"), std::string("test")}
                    }, std::string("Error occurred"), testError);
                }
                )->not->toThrow();
            }
            );
            it(std::string("should format error messages properly"), [=]() mutable
            {
                shared customLogger = createLogger();
                shared testError = std::make_shared<Error>(std::string("Test error"));
                testError->name = std::string("TestError");
                expect([=]() mutable
                {
                    return customLogger["error"](testError);
                }
                )->not->toThrow();
            }
            );
        }
        );
        describe(std::string("Clear Method"), [=]() mutable
        {
            it(std::string("should have clear method"), [=]() mutable
            {
                shared customLogger = createLogger();
                expect(type_of(customLogger["clear"]))->toBe(std::string("function"));
                expect([=]() mutable
                {
                    return customLogger["clear"]();
                }
                )->not->toThrow();
            }
            );
            it(std::string("should clear logs when called"), [=]() mutable
            {
                shared destination = const_((as<any>(logger)))[Symbol->for(std::string("pino-destination"))];
                expect(destination)->toBeDefined();
                expect(type_of(destination["clear"]))->toBe(std::string("function"));
                expect([=]() mutable
                {
                    return destination["clear"]();
                }
                )->not->toThrow();
            }
            );
        }
        );
        describe(std::string("Child Logger"), [=]() mutable
        {
            it(std::string("should create child logger with additional bindings"), [=]() mutable
            {
                auto customLogger = createLogger(object{
                    object::pair{std::string("parent"), std::string("main")}
                });
                auto childLogger = customLogger["child"](object{
                    object::pair{std::string("child"), std::string("sub")}
                });
                expect(childLogger)->toBeDefined();
                expect(type_of(childLogger["info"]))->toBe(std::string("function"));
            }
            );
        }
        );
        describe(std::string("Log Input Formats"), [=]() mutable
        {
            it(std::string("should handle various log input formats"), [=]() mutable
            {
                shared customLogger = createLogger();
                expect([=]() mutable
                {
                    return customLogger["info"](std::string("Simple string"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["info"](object{
                        object::pair{std::string("key"), std::string("value")}
                    }, std::string("With object"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["info"](std::string("Multiple"), std::string("string"), std::string("parts"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["error"](std::make_shared<Error>(std::string("Test")), std::string("With error"));
                }
                )->not->toThrow();
            }
            );
            it(std::string("should handle mixed arguments with errors"), [=]() mutable
            {
                shared customLogger = createLogger();
                shared error = std::make_shared<Error>(std::string("Test error"));
                expect([=]() mutable
                {
                    return customLogger["error"](std::string("Message"), error, object{
                        object::pair{std::string("extra"), std::string("data")}
                    });
                }
                )->not->toThrow();
            }
            );
        }
        );
        describe(std::string("Legacy Pino Compatibility"), [=]() mutable
        {
            it(std::string("should have legacy pino-compatible properties"), [=]() mutable
            {
                auto customLogger = createLogger();
                expect(customLogger["level"])->toBeDefined();
                expect(customLogger["levels"])->toBeDefined();
                expect(customLogger["levels"]["values"])->toBeDefined();
                expect(type_of(customLogger["isLevelEnabled"]))->toBe(std::string("function"));
                expect(type_of(customLogger["flush"]))->toBe(std::string("function"));
                expect(type_of(customLogger["flushSync"]))->toBe(std::string("function"));
            }
            );
            it(std::string("should have destination accessible via symbol"), [=]() mutable
            {
                auto customLogger = createLogger();
                auto destination = const_((as<any>(customLogger)))[Symbol->for(std::string("pino-destination"))];
                expect(destination)->toBeDefined();
                expect(type_of(destination["recentLogs"]))->toBe(std::string("function"));
                expect(type_of(destination["clear"]))->toBe(std::string("function"));
            }
            );
        }
        );
        describe(std::string("In-Memory Logger"), [=]() mutable
        {
            it(std::string("should store recent logs"), [=]() mutable
            {
                auto destination = const_((as<any>(logger)))[Symbol->for(std::string("pino-destination"))];
                expect(destination)->toBeDefined();
                auto logs = destination["recentLogs"]();
                expect(Array->isArray(logs))->toBe(true);
            }
            );
            it(std::string("should add logs with proper format"), [=]() mutable
            {
                auto destination = const_((as<any>(logger)))[Symbol->for(std::string("pino-destination"))];
                destination["addLog"](object{
                    object::pair{std::string("time"), Date->now()}, 
                    object::pair{std::string("level"), 30}, 
                    object::pair{std::string("msg"), std::string("Test message")}, 
                    object::pair{std::string("agentName"), std::string("TestAgent")}
                });
                auto logs = destination["recentLogs"]();
                auto lastLog = const_(logs)[logs["length"] - 1];
                expect(lastLog)->toBeDefined();
                expect(lastLog["msg"])->toBe(std::string("Test message"));
                expect(lastLog["agentName"])->toBe(std::string("TestAgent"));
            }
            );
        }
        );
        describe(std::string("Log Level Methods"), [=]() mutable
        {
            it(std::string("should support all log levels"), [=]() mutable
            {
                shared customLogger = createLogger();
                expect([=]() mutable
                {
                    return customLogger["fatal"](std::string("Fatal message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["error"](std::string("Error message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["warn"](std::string("Warn message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["info"](std::string("Info message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["debug"](std::string("Debug message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["trace"](std::string("Trace message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["verbose"](std::string("Verbose message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["log"](std::string("Log message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["progress"](std::string("Progress message"));
                }
                )->not->toThrow();
                expect([=]() mutable
                {
                    return customLogger["success"](std::string("Success message"));
                }
                )->not->toThrow();
            }
            );
        }
        );
        describe(std::string("Timestamp Configuration"), [=]() mutable
        {
            it(std::string("should respect LOG_TIMESTAMPS setting"), [=]() mutable
            {
                process->env->LOG_TIMESTAMPS = std::string("false");
                auto customLogger = createLogger();
                expect(customLogger)->toBeDefined();
            }
            );
            it(std::string("should default to showing timestamps"), [=]() mutable
            {
                process->env.Delete("LOG_TIMESTAMPS");
                auto customLogger = createLogger();
                expect(customLogger)->toBeDefined();
            }
            );
        }
        );
        describe(std::string("Sentry Integration"), [=]() mutable
        {
            it(std::string("should not capture exceptions when SENTRY_LOGGING is false"), [=]() mutable
            {
                process->env->SENTRY_LOGGING = std::string("false");
                shared customLogger = createLogger();
                shared error = std::make_shared<Error>(std::string("Test error"));
                expect([=]() mutable
                {
                    return customLogger["error"](error);
                }
                )->not->toThrow();
            }
            );
            it(std::string("should handle Sentry integration when enabled"), [=]() mutable
            {
                process->env->SENTRY_LOGGING = string_empty;
                shared customLogger = createLogger();
                shared error = std::make_shared<Error>(std::string("Sentry test"));
                expect([=]() mutable
                {
                    return customLogger["error"](error);
                }
                )->not->toThrow();
            }
            );
        }
        );
    }
    );
}

MAIN
